### PR TITLE
win32crypt improvements via #1859

### LINKED ISSUE #1859: win32crypt.CryptQueryObject not populating CertContext correctly when using AKV .cer file
(custom2.cer) When calling this function with a locally generated .cer file, it properly fully populates the dictionary as such
{'MsgAndCertEncodingType': 1, 'ContentType': 1, 'FormatType': 1, 'CertStore': <PyCERTSTORE object at 0x000002860D6F9E00>, 'Msg': None, 'Context': <PyCERT_CONTEXT object at 0x000002860D6F9DD0>}

(PEM_67ae84c1718440f4a89ace79a22e262d.cer) When a certificate is generated from Azure Key Vault and then downloaded as a .cer, the given output does not populate its context instead just having what is likely the pointer.
{'MsgAndCertEncodingType': 1, 'ContentType': 1, 'FormatType': 2, 'CertStore': <PyCERTSTORE object at 0x0000020ABC8C3A90>, 'Msg': None, 'Context': 2245136194336}

I have provided cert.py and the two certificates mentioned at https://teststoragebrklein.blob.core.windows.net/pywin32crypto?sp=r&st=2022-04-13T21:29:17Z&se=2022-05-01T05:29:17Z&spr=https&sv=2020-08-04&sr=c&sig=ySVJ82Fr0oE48wV8y0JyeQDeb52yWmWIYHdUkAueFDc%3D
Additionally wrote a simple C# solution also utilizing C interop which works fine which makes me think this is either a pywin32 or python issue.

* Version of Python and pywin32
Python 3.8.6
python -m pip freeze
pypiwin32==223
pywin32==303

### PR DESCRIPTION
**Pull request by mhammond**
This fixes #1859

### CODE REVIEW

**Comment by mhammond** [comment]
Hey @kxrob, if you can find the time, it would be great to have a second set of eyes on this to make sure I'm not doing anything dumb!

File: win32/src/win32crypt/win32cryptmodule.cpp [hunk 1]
```diff
@@ -1089,10 +1089,12 @@ static PyObject *PyCryptQueryObject(PyObject *self, PyObject *args, PyObject *kw
...
-    Py_BEGIN_ALLOW_THREADS bsuccess =
+    Py_BEGIN_ALLOW_THREADS;
+    bsuccess =
         CryptQueryObject(objecttype, input, contenttype, formattype, flags, &encoding, &contenttypeout, &formattypeout,
                          &hcertstore, &hcryptmsg, (const void **)&context);
-    Py_END_ALLOW_THREADS if (!bsuccess) return PyWin_SetAPIError("CryptQueryObject");
+    Py_END_ALLOW_THREADS;
+    if (!bsuccess) return PyWin_SetAPIError("CryptQueryObject");
 
     switch (formattypeout) {
```
Comment (kxrob) [inline win32/src/win32crypt/win32cryptmodule.cpp:1099]: This should probably be `switch (contenttypeout)`.  The issue #1859 bug - or reverse, the fact that it worked at all before with one of the cert files (the binary one), was that by chance CERT_QUERY_FORMAT_BINARY == CERT_QUERY_CONTENT_CERT numerically.  And CERT_QUERY_FORMAT_BASE64_ENCODED == CERT_QUERY_CONTENT_CTL caused the numeric context pointer result value - though the test file seems not to have a ctl.
Both cert test files seem to carry single certs only, one binary format, one base64  - no ctl.

File: win32/src/win32crypt/win32cryptmodule.cpp [hunk 2]
```diff
@@ -1101,8 +1103,7 @@ static PyObject *PyCryptQueryObject(PyObject *self, PyObject *args, PyObject *kw
             break;
         case CERT_QUERY_CONTENT_CTL:
         case CERT_QUERY_CONTENT_SERIALIZED_CTL:
-            // obcontext=new PyCTL_CONTEXT(context);
-            obcontext = PyLong_FromVoidPtr(context);
+            obcontext = PyWinObject_FromCTL_CONTEXT((PCCTL_CONTEXT)context);
```
Comment (kxrob) [inline win32/src/win32crypt/win32cryptmodule.cpp:1106]: (may work - test would require another test file with CTL)

File: win32/test/win32crypt_testcert_ctlcontext.cer [hunk 3]
```diff
@@ -0,0 +1 @@
+MIIDKjCCAhKgAwIBAgIQaywn/6CWRlGT/OAEKRt05TANBgkqhkiG9w0BAQsFADASMRAwDgYDVQQDEwdicmFuZG9uMB4XDTIyMDQxMjE1MjQyOFoXDTIzMDQxMjE1MzQyOFowEjEQMA4GA1UEAxMHYnJhbmRvbjCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAMTXRWxv/xyYMy7BIUj6obtQ9+prBV+6dgmvIXMV5OqirGb2ldA3lP2yO1VSnxchFSLuwGNsEc9R6MYwa+ftPuJMQETzCpQTDZqmu8LqZAxkqU6D4sqHWU3DhSTE0M1vBBTFN8vSvrhAAnH0EJ1sEfIHsxOMzksEMWxFVzR6yMwgTZe1akPq5AfthSdKmN+XI3nIliiJbeq1SpFvNrj5tpdp3xqr+G4wjuBzghgmMAE4C/xAQ9b43VAbQvQiw5JSFU9sAQeUvFdfDLg/tVN5aDmJUdVLQ9oXMAJv1ZfJGdpPwEsGGhnHx0Qrvpn1jh56Nq7tDhsalgkoJw2HhPfVwvUCAwEAAaN8MHowDgYDVR0PAQH/BAQDAgWgMAkGA1UdEwQCMAAwHQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCMB8GA1UdIwQYMBaAFPHRgQjh+QpSvNiXHdqWbSwIquZ7MB0GA1UdDgQWBBTx0YEI4fkKUrzYlx3alm0sCKrmezANBgkqhkiG9w0BAQsFAAOCAQEAIvLp6BqDiI9Hh40b1+EOIJetX0pf6gyi16N/s7d+fR3hHxerfLxJ4tB2FR1YCx+xf1h66ucaPEBHaokOyQ+UOuLcoc0KHOv+MIzZ5G8d7lv7AGpPn3F0OwwRuo4SxJNgxRYYynOfQf5QTs5rJkGBmMRehM6tSIlZWymFvsBxxCjI+SIiRa7/iL1TEUDMbMQI6MG0i8uLrOHGO7ioChk+rgXEH82gl6nPQx8pQrVTzv2NAJXVPWNhrGci/64b0OqJ4WFUrqbPXH7Vp27ROUD4PfB9CgVxBA+wjDwHeOAMmKxUvA1jGwyqf0VE11mJaSxH2luNv7/z+dGSUQDEXCqWlw==
```
Comment (kxrob) [inline win32/test/win32crypt_testcert_ctlcontext.cer:1]:
```
...
DBTE>>> cert['ContentType']  
1      # CERT_QUERY_CONTENT_CERT  The content is a single certificate. 
DBTE>>> cs1=cert['CertStore']
DBTE>>> cs1.CertEnumCTLsInStore()
[]
DBTE>>> cs1.CertEnumCertificatesInStore()
[<PyCERT_CONTEXT object at 0x0000021613956B90>]
```

=> This file seems not to have a CTL context but only a single cert in base64 format

File: win32/Lib/pywin32_testutil.py [hunk 4]
```diff
@@ -193,6 +194,24 @@ def check_is_admin():
...
+        os.path.dirname(sys.argv[0]),
+        extra_dir,
+        ".",
+    ]
+    for candidate in candidates:
+        fname = os.path.join(candidate, basename)
+        if os.path.isfile(fname):
+            return fname
+    else:
+        raise TestSkipped(f"Can't find test fixture '{fname}'")
```
Comment (kxrob) [inline win32/Lib/pywin32_testutil.py:212]: I wonder if silent test-skipping is the appropriate (default) action when file not found? The cert files at least are in the repo and always should be found when no coding error?

Current CI test runs show silent skip: "SKIPPED: 1 tests - Can't find test fixture '.\01-Intro.wav'"
Is this acceptable / intended / file erroneously missing / filename or folder or code wrong ... ?
If skipping is acceptable for some use-cases by exception, maybe optional `missing_skips=True` arg,  or None return + specific handling in the caller ... ?

(It seems in the initial PR push the CI cert tests also seemed to be skipped silently when the cert files were not found, which is probably not acceptable.)

**Review by kxrob — COMMENTED** [review]
(see code review comments)

File: win32/Lib/pywin32_testutil.py (hunk 4 above)
Comment (mhammond) [reply win32/Lib/pywin32_testutil.py:212]: great question :) I opened https://github.com/mhammond/pywin32/pull/1880! This patch worked around this by installing the .cert files, which isn't necessary imo - so I've killed alot of that.

File: win32/src/win32crypt/win32cryptmodule.cpp (hunk 1 above)
Comment (mhammond) [reply win32/src/win32crypt/win32cryptmodule.cpp:1099]: ouch - thank you! I really should have noticed that was a base64 file 🤦 It's quite odd that it actually "worked" - I did get a PyCTL_CONTEXT from that certificate, and managed to use it as such. But you are clearly correct, thanks. Sadly I'm back to no coverage for PyCTL_CONTEXT but I think I'll leave the changes around that object in-place.

**Comment by mhammond** [comment]
Thanks for that - I'll get a new PR up soon

File: win32/Lib/pywin32_testutil.py (hunk 4 above)
Comment (kxrob) [reply win32/Lib/pywin32_testutil.py]: There is still 
```
SKIPPED: 2 tests - Can't find test fixture '.\win32crypt_testcert_base64.cer'
SKIPPED: 1 tests - Can't find test fixture '.\win32crypt_testcert_bin.cer'
```
in the last logs here so far.
But after rebasing beyond d90ad11 it probably works - as happened with  '.\01-Intro.wav' ...

Comment (kxrob) [reply win32/Lib/pywin32_testutil.py]: Yes does work in CI run of a test rebase (conflict resolution was only in CHANGES.txt):
https://github.com/kxrob/pywin32/commit/6e3a059e9ac5c81580f0ec940894ebf9cd63bc8a

**Pull request merged**